packages/squiggle-lang/src/rescript/Utility/E/E_A.res

```diff
@@ -323,32 +323,45 @@ module Floats = {
       let continuous: array<float> = []
       let discrete = FloatFloatMap.empty()
 
-      let addData = (count: int, value: float): unit => {
-        if count >= minDiscreteWeight {
-          FloatFloatMap.add(value, count->Belt.Int.toFloat, discrete)
+      // In a run of exactly minDiscreteWeigth, the first and last
+      // element indices differ by minDistance.
+      let minDistance = minDiscreteWeight - 1
+
+      let l = length(sortedArray)
+      let i = ref(0)
+      while i.contents < l - minDistance {
+        let value = sortedArray[i.contents]
+        if value != sortedArray[i.contents + minDistance] {
+          Js.Array2.push(continuous, value) |> ignore
+          i := i.contents + 1
         } else {
-          for _ in 1 to count {
-            continuous->Js.Array2.push(value)->ignore
+          // Move i forward to next unequal value
+          // That is, find j so that equalAt(j-1) and !equalAt(j)
+          let i0 = i.contents
+          // Find base so that j is in (i0+base, i0+2*base]
+          let base = ref(minDistance)
+          let equalAt = (ind: int) => ind < l && sortedArray[ind] == value
+          while equalAt(i0 + base.contents * 2) {
+            base := base.contents * 2
           }
-        }
-      }
-
-      let (finalCount, finalValue) = sortedArray->Belt.Array.reduce(
-        // initial prev value doesn't matter; if it collides with the first element of the array, flush won't do anything
-        (0, 0.),
-        ((count, prev), element) => {
-          if element == prev {
-            (count + 1, prev)
-          } else {
-            // new value, process previous ones
-            addData(count, prev)
-            (1, element)
+          // Maintain j in (lo, i]. Once lo+1 == i, i is j.
+          let lo = ref(i0 + base.contents)
+          i := Js.Math.min_int(lo.contents + base.contents, l)
+          while lo.contents + 1 < i.contents {
+            let mid = lo.contents + (i.contents - lo.contents) / 2
```

Comment:
This is a useful trick when there's a chance of integer overflow, but we'll never have sample sets in 1e9 range, so `(lo.contents + i.contents) / 2` would be more readable.

Reply:
I'd rather not leave that sort of dangerous code lying around regardless. Also the base/range representation makes various other things a lot easier, so I prefer to make that form obvious in case I use this code for later reference on something else.